src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1354,19 +1357,90 @@ public ValidateCountAttribute(int minLength, int maxLength)
+    /// </summary>
+    public abstract class CachedValidValuesGeneratorBase : IValidateSetValuesGenerator
+    {
+        // Cached valid values.
+        private string[] _validValues;
+
+        /// <summary>
+        /// Initializes a new instance of the CachedValidValuesGeneratorBase class.
+        /// </summary>
+        /// <param name="cacheExpirationInSeconds">Sets a time interval in seconds to reset the '_validValues' dynamic valid values cache.</param>
+        public CachedValidValuesGeneratorBase(int cacheExpirationInSeconds)
+        {
+            _validValuesCacheExpiration = cacheExpirationInSeconds;
+        }
+
+        private int _validValuesCacheExpiration;
+
+        /// <summary>
+        /// Abstract method to generate a valid values.
+        /// </summary>
+        public abstract string[] GenerateValidValues();
+
+        /// <summary>
+        /// Get a valid values.
+        /// </summary>
+        public string[] GetValidValues()
+        {
+            // Because we have a background task to clear the cache by '_validValues = null'
+            // we use the local variable to exclude a race condition.
+            var validValuesLocal = _validValues;
+            if (validValuesLocal != null)
+            {
+                return validValuesLocal;
+            }
+
+            var validValuesNoCache = GenerateValidValues();
+
+            if (validValuesNoCache == null)
+            {
+                throw new ValidationMetadataException(
+                    "ValidateSetGeneratedValidValuesListIsNull",
+                    null,
+                    Metadata.ValidateSetGeneratedValidValuesListIsNull);
+            }
+
+            if (_validValuesCacheExpiration > 0)
+            {
+                Task.Delay(_validValuesCacheExpiration * 1000).ContinueWith((task) => _validValues = null);
+                _validValues = validValuesNoCache;
+            }
+
+            return validValuesNoCache;
+        }
+    }
+
+    /// <summary>
     /// Validates that each parameter argument is present in a specified set
     /// </summary>
     [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
     public sealed class ValidateSetAttribute : ValidateEnumeratedArgumentsAttribute
     {
+        // We can use either static '_validValues'
+        // or dynamic valid values list generated by instance of 'validValuesGenerator'.
         private string[] _validValues;
+        private IValidateSetValuesGenerator validValuesGenerator = null;
+
+        // The valid values generator cache works across 'ValidateSetAttribute' instances.
+        private static ConcurrentDictionary<Type, IValidateSetValuesGenerator> s_ValidValuesGeneratorCache = new ConcurrentDictionary<Type, IValidateSetValuesGenerator>();
+
+        /// <summary>
+        /// Clear the valid values generator cache.
+        /// </summary>
+        public static void ClearValidValuesGeneratorCache()
```

Comment:
I'm not sure this is useful - because each instance of the attribute will also contain a reference to the `IValidateSetValuesGenerator` and we can't clear that.

When I suggested a method to clear the cache, I was referring to `CachedValidValuesGeneratorBase`, but I think it's probably fine to not add that - one can simply use a smaller timeout value or implement their own custom caching logic.